### PR TITLE
Initialize data buffer when neural network initializes.

### PR DESCRIPTION
**Pull request by jijoongmoon**
Initilaize data buffer and set parameters from neural network.

Signed-off-by: jijoongmoon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/include/databuffer.h [hunk 1]
```diff
@@ -215,6 +248,7 @@ class DataBuffer {
   std::string val_file;
   std::string test_file;
   std::vector<std::string> labels;
+  bool validation[4];
```
Comment (zhoonit) [inline nntrainer/include/databuffer.h:251]: Looks like it seems to be `NBUFTYPE`
BTW, could I ask what `validation` is for?

Comment (jijoongmoon) [reply nntrainer/include/databuffer.h:251]: right. Thanks.

File: nntrainer/src/databuffer.cpp [hunk 2]
```diff
@@ -106,6 +116,53 @@ bool DataBuffer::init(int mini_batch, unsigned int train_bufsize,
...
 }
 
+int DataBuffer::init() {
+
+  int status = ML_ERROR_NONE;
+
+  if (!this->class_num) {
+    ml_loge("Error: number of class must be set");
+    for (int i = 0; i < NBUFTYPE; ++i)
+      validation[i] = false;
```
Comment (dongju-chae) [inline nntrainer/src/databuffer.cpp:126]: Q: I wonder how this per-buffer validation value will be utilized. I think a single `initialized` variable may perform similar behaviors.

File: nntrainer/src/databuffer.cpp [hunk 3]
```diff
@@ -106,6 +116,53 @@ bool DataBuffer::init(int mini_batch, unsigned int train_bufsize,
...
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  this->rest_train = max_train;
+  this->rest_val = max_val;
+  this->rest_test = max_test;
+
+  this->train_running = true;
```
Comment (helloahn) [inline nntrainer/src/databuffer.cpp:156]: Q) As far as I know, carry out each step in the order of the train-validation->train->validation...->test(at last). Is it okay if all the step running flags are made true?

File: nntrainer/src/databuffer.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:126]: There will be three cases (Training, Validation, Testing) with Label data. 
validation[0]: Training Set
validation[1]: Validation Set
validation[2]: Testing Set
validation[3]: Label data

Training and label is essential and others are optional. 
Depending on this data provided,  consequence procedure will be coducted.

File: nntrainer/src/databuffer.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:156]: This is the flag to synchronize with data generation thread. If this flag is false, then the data generation thread will stop there.

Comment (helloahn) [reply nntrainer/src/databuffer.cpp:156]: I got it thanks for the explanation!

File: nntrainer/src/databuffer.cpp (hunk 2 above)
Comment (dongju-chae) [reply nntrainer/src/databuffer.cpp:126]: Ah, some buffers would be optional. Got it.
